mods/logon-logoff-shutdown-sounds.wh.cpp

```diff
@@ -0,0 +1,582 @@
+            break;
+        case ST_LOGOFF:
+            szSoundType = TEXT("WindowsLogoff");
+            break;
+        case ST_EXIT:
+            szSoundType = TEXT("SystemExit");
+            break;
+        case ST_START:
+            szSoundType = TEXT("SystemStart");
+            break;
+    }
+    
+    pcbData[0] = 520;
+    
+    hr = wsprintf(szSubKey, TEXT("AppEvents\\Schemes\\Apps\\.Default\\%ws\\.Current"), szSoundType);
+
+    if (SUCCEEDED(hr))
+    {
+        hr = RegGetValue(HKEY_CURRENT_USER, szSubKey, 0, 2u, 0, pvData, pcbData);
+        if (hr == S_OK)
+            return PlaySoundFile(hThread, pvData);
+        else
+            return HRESULT_FROM_WIN32((USHORT)hr);
+    }
+    return hr;
+}
+
+LRESULT SHDefWindowProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
+{
+    if (IsWindowUnicode(hwnd))
+    {
+        return DefWindowProcW(hwnd, uMsg, wParam, lParam);
+    }
+    else
+    {
+        return DefWindowProcA(hwnd, uMsg, wParam, lParam);
+    }
+}
+
+STDAPI_(BOOL) SHRegisterClassW(const WNDCLASSW* pwc)
+{
+    WNDCLASSW wc;
+    if (!GetClassInfo(pwc->hInstance, pwc->lpszClassName, &wc)) 
+    {
+        return RegisterClass(pwc);
+    }
+    return TRUE;
+}
+
+HWND SHCreateWorkerWindowW(WNDPROC pfnWndProc, HWND hwndParent, DWORD dwExStyle, DWORD dwFlags, HMENU hmenu, void * p)
+{
+    WNDCLASSW wc = {0};
+
+    wc.lpfnWndProc      = DefWindowProcW;
+    wc.cbWndExtra       = sizeof(void *);
+    wc.hInstance        = g_hShlwapi;
+    wc.hCursor          = LoadCursor(NULL, IDC_ARROW);
+    wc.hbrBackground    = (HBRUSH) (COLOR_BTNFACE + 1);
+    wc.lpszClassName    = L"WorkerW";
+    //dwExStyle |= IsBiDiLocalizedSystem() ? /*dwExStyleRTLMirrorWnd*/0L : 0L;
+
+    SHRegisterClassW(&wc);
+
+    HWND hwnd = CreateWindowExW(dwExStyle, L"WorkerW", NULL, dwFlags,
+                                  0, 0, 0, 0, hwndParent,
+                                  (HMENU)hmenu, g_hShlwapi, NULL);
+    if (hwnd) 
+    {
+        SetWindowLongPtr(hwnd, 0, (LPARAM)(p));
+
+        // Note: Must explicitly use W version to avoid charset thunks
+        if (pfnWndProc)
+            SetWindowLongPtrW(hwnd, GWLP_WNDPROC, (LONG_PTR)pfnWndProc);
+
+    }
+
+    return hwnd;
+}
+
+class CSoundWnd {
+    public:
+        CSoundWnd();
+        LONG m_refCount;
+        HWND m_hwndSound;
+        HANDLE m_thread;
+        BOOL __thiscall Init();
+        static DWORD CALLBACK s_ThreadProc(void* lpParam);
+        LONG __thiscall Release();
+        static DWORD CALLBACK s_CreateWindow(void* lpParam);
+        static LRESULT CALLBACK s_WndProc(HWND hWnd,
+                                   UINT msg,
+                                   WPARAM wParam,
+                                   LPARAM lParam);
+        LRESULT v_WndProc(HWND hWnd, UINT mst, WPARAM wParam, LPARAM lParam);
+};
+
+CSoundWnd::CSoundWnd()
+    : m_refCount(1)
+    , m_hwndSound(nullptr)
+    , m_thread(nullptr)
+{
+}
+
+BOOL CSoundWnd::Init()
```

Comment:
I see that some of the code is based on code from ExplorerPatcher. If significant parts are copied or heavily inspired by other projects, consider crediting them and specifying a compatible license for your mod.